internal/node/controller/auth.go

```diff
@@ -103,3 +109,186 @@ func getUserCertAndInfo(controller NodeController, username string) (*x509.Certi
+	}
+}
+
+func (p *BFFProvider) handleChallenge(w http.ResponseWriter, r *http.Request) {
+	var req BFFChallengeRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, "Invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	_, ok := p.friends[req.DID]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Friend with DID %s not found", req.DID), http.StatusNotFound)
+		return
+	}
+
+	challenge, err := bffauth.GenerateChallenge()
+	if err != nil {
+		http.Error(w, "Failed to generate challenge", http.StatusInternalServerError)
+		return
+	}
+
+	// Create session with UUID
+	sessionID := uuid.New().String()
+	session := &bffauth.ChallengeSession{
+		SessionID: sessionID,
+		DID:       req.DID,
+		Challenge: challenge,
+		ExpiresAt: time.Now().Add(15 * time.Minute),
+	}
+	err = p.challengePersister.SaveSession(session)
+	if err != nil {
+		http.Error(w, "Failed to save session", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(w).Encode(map[string]string{
+		"challenge": challenge,
+		"session":   sessionID,
+	})
+	if err != nil {
+		http.Error(w, "Failed to encode response", http.StatusInternalServerError)
+		return
+	}
+}
+
+func (p *BFFProvider) handleAuth(w http.ResponseWriter, r *http.Request) {
+	var req struct {
+		SessionID string `json:"session_id"`
+		Proof     string `json:"proof"`
+	}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		http.Error(w, "Invalid request body", http.StatusBadRequest)
+		return
+	}
+
+	session, err := p.challengePersister.GetSession(req.SessionID)
+	if err != nil {
+		http.Error(w, "Invalid session", http.StatusBadRequest)
+		return
+	}
+
+	err = p.challengePersister.DeleteSession(req.SessionID)
+	if err != nil {
+		http.Error(w, "Failed to delete session", http.StatusInternalServerError)
+		return
+	}
+
+	friend, ok := p.friends[session.DID]
+	if !ok {
+		http.Error(w, fmt.Sprintf("Friend with DID %s not found", session.DID), http.StatusNotFound)
+		return
+	}
+
+	// Verify proof
+	valid, err := bffauth.VerifyProof(session.Challenge, req.Proof, friend.PublicKey)
+	if err != nil || !valid {
+		http.Error(w, "Invalid proof", http.StatusUnauthorized)
+		return
+	}
+
+	// Generate JWT
+	token, err := bffauth.GenerateJWT(p.signingKey)
+	if err != nil {
+		http.Error(w, "Failed to generate token", http.StatusInternalServerError)
+		return
+	}
+
+	// Sanity check by validating the token
+	_, err = bffauth.ValidateJWT(token, p.signingKey)
+	if err != nil {
+		http.Error(w, "Failed to validate token", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(w).Encode(map[string]string{
+		"token": token,
+	})
+	if err != nil {
+		http.Error(w, "Failed to encode response", http.StatusInternalServerError)
+		return
+	}
+}
+
+func (p *BFFProvider) handleAddFriend(w http.ResponseWriter, r *http.Request) {
```

Comment:
can you explain in a little more detail why we need to do this first (i guess i'm looking for some minimal documentation on the protocol)